docs/site/Testing-your-application.md

```diff
@@ -408,17 +408,68 @@ Verify how was the stubbed method executed at the end of your unit test (in the
 ```ts
 // expect that repository.find() was called with the first
 // argument deeply-equal to the provided object
-sinon.assert.calledWithMatch({where: {id: 1}});
+sinon.assert.calledWithMatch(findStub, {where: {id: 1}});
 ```
 
 See [Unit test your controllers](#unit-test-your-controllers) for a full
 example.
 
 #### Create a stub Service
 
-{% include content/tbd.html %}
+If your controller relies on service proxy for service oriented backends such as
+REST APIs, SOAP Web Services, or gRPC microservices, then we can create stubs
+for the service akin to the steps outlined in the above
+[stub repository](#create-a-stub-repository) section. Consider a dependency on a
+`GeoCoder` service that relies on a remote REST API for returning coordinates
+for a specific address.
 
-The initial beta release does not include Services as a first-class feature.
+```ts
+class DummyGeocoder implements GeocoderService {
+  // in practice, the constructor could take an API key to authenticate with
+  // the service proxy provider
+  constructor() {}
+
+  geocode(address: string) {
+    // this would call the remote API for the real coordinates
+  }
+}
+```
+
+Create a test double for the service class:
+
+```ts
+describe('GeoCoderController', () => {
+  let geoService: DummyGeocoder;
+  beforeEach(givenStubbedService);
+
+  // your unit tests
+
+  function givenStubbedService() {
+    geoService = sinon.createStubInstance(DummyGeocoder);
+  }
+});
```

Comment:
Your example is showing a different approach from what we have in our Todo example. I find it impractical that we have to create a `DummyGeocoder` class, when all we want is a mocked instance with `GeocoderService` API.

Here is what we do in our example - see https://github.com/strongloop/loopback-next/blob/0cccae6e2a7a9eb08feee6ab57730e8c10e4269c/examples/todo/src/__tests__/unit/controllers/todo.controller.unit.ts

```ts
describe('TodoController', () => {
  let geoService: GeocoderService;
  let geocode: sinon.SinonStub;
 
  beforeEach(resetRepositories);

  // (the tests)

  function resetRepositories() {
    // ...

    // this creates a stub with GeocoderService API
    // in a way allowing the compiler to verify type correctness
    geoService = {geocode: sinon.stub()};

    // this creates a reference to the stubbed "geocode" method
    // because "geoService.gecode" has type from GeocoderService
    // and does not provide Sinon APIs
    geocode = geoService.geocode as sinon.SinonStub;

    controller = new TodoController(todoRepo, geoService);
  }
});
```

Having said that, your approach has its benefits too. If we use `createStubInstance` from testlab instead of `sinon.createStubInstance`, then we don't have to cast stubbed methods to `sinon.SinonStub`.

```ts
let geoService: StubbedInstanceWithSinonAccessor<GeocoderService>;
// ...
geoService = createStubInstance(DummyGeocoder);
// ...
geoService.stubs.geocode.resolves([<GeoPoint>{y: 41.109653, x: -73.72467}]);
```

See how TodoController tests are stubbing TodoRepository and its methods.

Reply:
Yea I looked at those examples (and https://github.com/strongloop/loopback-next/blob/master/packages/service-proxy/src/__tests__/acceptance/service.mixin.acceptance.ts#L18-L28) and decided to make it similar to how we were stubbing repositories. I don't mind changing it back to a mocked service of the `GeoService` API, especially since I'm pointing back to the test case in our Todo example.

Reply:
I think my main concern is about consistency between the code in our examples and the documentation. If you prefer the approach proposed in this PR, then feel free to go with it, just update the examples to use it too.